lib/Transform/XTenQDQConcat.cpp

```diff
@@ -0,0 +1,67 @@
+#include "xten/Dialect/XTenNN/IR/XTenNNOps.h"
+#include "xten/Transform/Passes.h"
+
+#include <mlir/Dialect/Tosa/IR/TosaOps.h>
+#include <mlir/IR/Matchers.h>
+#include <mlir/IR/PatternMatch.h>
+#include <mlir/Support/LogicalResult.h>
+
+#include <llvm/ADT/SmallVector.h>
+
+using namespace mlir;
+using namespace mlir::tosa;
+using namespace amd::xten_nn;
+
+namespace {
+struct XTenQDQConcat : public OpRewritePattern<DequantizeOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(DequantizeOp op,
+                                PatternRewriter &rewriter) const override {
+    // Match concat->QDQ->concat and remove QDQ, if concats would be foldable.
+    // Try to match an incoming concat
+    Value val;
+    if (!matchPattern(op.getInput(), m_Op<QuantizeOp>(matchers::m_Any(&val)))) {
+      return rewriter.notifyMatchFailure(
+          op, "Dequantize input not produced by Quantize.");
+    }
```

Comment:
To be a bit less destructive, should we check that quantize and dequantize have the same scale?

Reply:
That's actually a good check. Will do